From 3c09f7542ebe0a50b1f869db1e59a7c4f4fef006 Mon Sep 17 00:00:00 2001
From: Dennis Marttinen <twelho@welho.tech>
Date: Wed, 25 Mar 2026 19:29:46 +0200
Subject: [PATCH] md/bcache: eliminate blocking empty flushes against backing
 device

In writeback mode, empty writes and flush requests still issue flushes against the backing device and wait for it to complete. This is disastrous for performance with a PLP NVMe, as it degrades the instant NVMe flushes to HDD seeks.
---
 drivers/md/bcache/request.c | 24 ------------------------
 1 file changed, 24 deletions(-)

diff --git a/drivers/md/bcache/request.c b/drivers/md/bcache/request.c
index 0aa9d623f00dbd..c30427eb832614 100644
--- a/drivers/md/bcache/request.c
+++ b/drivers/md/bcache/request.c
@@ -1029,26 +1029,6 @@ static void cached_dev_write(struct cached_dev *dc, struct search *s)
 	} else if (s->iop.writeback) {
 		bch_writeback_add(dc);
 		s->iop.bio = bio;
-
-		if (bio->bi_opf & REQ_PREFLUSH) {
-			/*
-			 * Also need to send a flush to the backing
-			 * device.
-			 */
-			struct bio *flush;
-
-			flush = bio_alloc_bioset(bio->bi_bdev, 0,
-						 REQ_OP_WRITE | REQ_PREFLUSH,
-						 GFP_NOIO, &dc->disk.bio_split);
-			if (!flush) {
-				s->iop.status = BLK_STS_RESOURCE;
-				goto insert_data;
-			}
-			flush->bi_end_io = backing_request_endio;
-			flush->bi_private = cl;
-			/* I/O request sent to backing device */
-			closure_bio_submit(s->iop.c, flush, cl);
-		}
 	} else {
 		s->iop.bio = bio_alloc_clone(bio->bi_bdev, bio, GFP_NOIO,
 					     &dc->disk.bio_split);
@@ -1070,10 +1050,6 @@ static CLOSURE_CALLBACK(cached_dev_nodata)
 	if (s->iop.flush_journal)
 		bch_journal_meta(s->iop.c, cl);
 
-	/* If it's a flush, we send the flush to the backing device too */
-	bio->bi_end_io = backing_request_endio;
-	closure_bio_submit(s->iop.c, bio, cl);
-
 	continue_at(cl, cached_dev_bio_complete, NULL);
 }
 
